### ISSUE TITLE
Content-Length Header on Downloads

### ISSUE THREAD

**Issue by subins2000** [owncloud/core#24520]
### Steps to reproduce
1. Make a file shareable
2. Go to the [file's public page](https://sky-phpgeek.rhcloud.com/s/VtQfWaP2FQUNFNG)
3. [Download File](https://sky-phpgeek.rhcloud.com/s/VtQfWaP2FQUNFNG/download)

### Expected behaviour
The file should download with "Content-Length" header.

### Actual behaviour
The "Content-Length" header is omitted by ownCloud.

Test - http://goo.gl/CJizER

### Problem

Due to this, browser won't show the file size and time remaining to complete download.

Also, connection is closed before file is downloaded completely.

### Server configuration
**Operating system**: Redhat Linux (OpenShift Cloud)

**Web server:** Apache

**Database:** MySQL

**PHP version:** 5.5

**ownCloud version:** 9.0.2

**Updated from an older ownCloud or fresh install:** fresh install

**Where did you install ownCloud from:** http://owncloud.org/install

**Signing status (ownCloud 9.0 and above):** 

```
Technical information
=====================
The following list covers which files have failed the integrity check. Please read
the previous linked documentation to learn more about the errors and how to fix
them.

Results
=======
- core
	- FILE_MISSING
		- 3rdparty/doctrine/dbal/bin/doctrine-dbal
		- 3rdparty/doctrine/dbal/bin/doctrine-dbal.php
		- 3rdparty/react/promise/.gitignore
		- 3rdparty/react/promise/CHANGELOG.md
		- 3rdparty/sabre/uri/tests/BuildTest.php
		- 3rdparty/sabre/uri/tests/NormalizeTest.php
		- 3rdparty/sabre/uri/tests/ResolveTest.php
		- 3rdparty/sabre/uri/tests/SplitTest.php
		- 3rdparty/sabre/uri/tests/phpcs/ruleset.xml
		- 3rdparty/sabre/uri/tests/phpunit.xml.dist
		- core/vendor/.gitignore
		- core/vendor/davclient.js/lib/client.js
		- core/vendor/es6-promise/.bower.json
		- core/vendor/es6-promise/.npmignore
		- core/vendor/es6-promise/.release.json
		- core/vendor/es6-promise/dist/es6-promise.js

Raw output
==========
Array
(
    [core] => Array
        (
            [FILE_MISSING] => Array
                (
                    [3rdparty/doctrine/dbal/bin/doctrine-dbal] => Array
                        (
                            [expected] => 3ea01094f0ec11aab8b77ab25203703790eebda7b64afa971f678597552fb29789258f648df21d17591504f531ea54bc276830c8e3e83b548f89967b9ed1987c
                            [current] => 
                        )

                    [3rdparty/doctrine/dbal/bin/doctrine-dbal.php] => Array
                        (
                            [expected] => 8012a9dc5670451456ba5d4c77ab3865e9f9020edab88560f8f964ead1f8e141885dd7c6f23afa66fe865337a0602d330c822467a5ca3dcc95ed85bb46aaa10d
                            [current] => 
                        )

                    [3rdparty/react/promise/.gitignore] => Array
                        (
                            [expected] => 3fbb564fd8c8e725d0227f7624fa78614235b5fa399ed68fa6ff61b12a3bc7b11fd96ae34b3be375222906b791e43bba072977939a80c488806df3e0937d678a
                            [current] => 
                        )

                    [3rdparty/react/promise/CHANGELOG.md] => Array
                        (
                            [expected] => 8a38e5b9135f6335a0b992c4501c51a1e9ffe52c94b1f301790f870f5b4a812e08788591b1d58d793d8611a9670d272d1eb9260888304aef0c53e1467dce117b
                            [current] => 
                        )

                    [3rdparty/sabre/uri/tests/BuildTest.php] => Array
                        (
                            [expected] => 972eca809a46e45fd1fd54d9010efea9a606ec2c3fb8ee8c98e2e8564cbe7139153d40906ebef009859d112e772fa8c11c17ea74abb9b2c03a04ddc884533ec5
                            [current] => 
                        )

                    [3rdparty/sabre/uri/tests/NormalizeTest.php] => Array
                        (
                            [expected] => 652a24da4a4755e054c4c53e0335d823be9f9d44a05e43549b91873571bfd6659f0c141e30a394b3fa208543e89f2b6b020103c8c7117dceb65fe718a207e14e
                            [current] => 
                        )

                    [3rdparty/sabre/uri/tests/ResolveTest.php] => Array
                        (
                            [expected] => 98923f3be5261991c7fd993a0312def5923719b72ebe13c1d3fc4aa8098398a6e18eb8252cf1563a4a9b1b7a85869805d21e94b2e703447e8911d7bd759259e9
                            [current] => 
                        )

                    [3rdparty/sabre/uri/tests/SplitTest.php] => Array
                        (
                            [expected] => 7c4dd4aa3dbff7878d4577d6b9140459e063f5c9b969d40dd4b1a9fbe808b40a8b2448d6da3345af96c5aaac406a35e2d4b4247c841f6d680434cbecead80fc0
                            [current] => 
                        )

                    [3rdparty/sabre/uri/tests/phpcs/ruleset.xml] => Array
                        (
                            [expected] => c28690c26c63f03454a10d81b184dd55a9d34fcf46fca12d0eb1fa59356341903c8e448d851927a7189aaeaae532d3750c959a6000770af685434ba2968e56a4
                            [current] => 
                        )

                    [3rdparty/sabre/uri/tests/phpunit.xml.dist] => Array
                        (
                            [expected] => c4af0b485e9c7a2089e7da37af217b414ab4187cd592e8de726aec316b8c061b91456eee927b74dc4795bd0037f346587b61f3bec285f04086c3d7e5e50409f5
                            [current] => 
                        )

                    [core/vendor/.gitignore] => Array
                        (
                            [expected] => 0585b06955afefe4990028cf29062c30145210b3f199359207e4a915f178513fcdf4269106a779ce7b0f0d16bd57fb97a2ed5fdb85813d7f1ff2368e51b0cbd4
                            [current] => 
                        )

                    [core/vendor/davclient.js/lib/client.js] => Array
                        (
                            [expected] => 81d26d67f23736ded76e7ecadd0eb2040820fe65c1c9f57cb348be427d6373c215ecf16fde42a4684d095e01ffd27b14c0342966939896fda2dcad71a87d99cb
                            [current] => 
                        )

                    [core/vendor/es6-promise/.bower.json] => Array
                        (
                            [expected] => 170374d66abd2edf763541fe8324ee89530bebad469c6c322a5aa50f4b8562d66196c1fd25732841c9090e09c40e29a73202eb30f3375fffaddf762c948e0d9e
                            [current] => 
                        )

                    [core/vendor/es6-promise/.npmignore] => Array
                        (
                            [expected] => ee3d85b16688b6eb25a68684f3eb68dce6fea84a6b39823336c36770c15fbdf32961de908a72b210a2d409578c9171b9131300511b1804443f4844028482fac7
                            [current] => 
                        )

                    [core/vendor/es6-promise/.release.json] => Array
                        (
                            [expected] => 6d62b3c3d7cc97aba9077a99ad63d4423d4af243c39bb62f26049429d4c9b7c25ec81dd9693e85be9632729ce34c62608f1eda32e50608aecfe59d4fb03c2d20
                            [current] => 
                        )

                    [core/vendor/es6-promise/dist/es6-promise.js] => Array
                        (
                            [expected] => 4a73730204beef284b317555fec1a3a690723a84cd43b85bbd70e6feb7a51206927633f0a6c6a4762210decd6513cac31b311ffeac94c24b915eb4b1ddcfea3f
                            [current] => 
                        )

                )

        )

)
```

**List of activated apps:**
```
Enabled:
  - activity: 2.2.1
  - comments: 0.2
  - dav: 0.1.6
  - federatedfilesharing: 0.1.0
  - federation: 0.0.4
  - files: 1.4.4
  - files_pdfviewer: 0.8.1
  - files_sharing: 0.9.1
  - files_texteditor: 2.1
  - files_trashbin: 0.8.0
  - files_versions: 1.2.0
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - gallery: 14.5.0
  - notifications: 0.2.3
  - provisioning_api: 0.4.1
  - systemtags: 0.2
  - templateeditor: 0.1
  - updatenotification: 0.1.0
Disabled:
  - encryption
  - external
  - files_external
  - user_external
  - user_ldap
```

**The content of config/config.php:**

```
{
    "system": {
        "instanceid": "ocdau4bxrjyv",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "sky-phpgeek.rhcloud.com"
        ],
        "datadirectory": "\/var\/lib\/openshift\/572f7fb30c1e6688f2000219\/app-root\/data\/",
        "overwrite.cli.url": "https:\/\/sky-phpgeek.rhcloud.com",
        "dbtype": "mysql",
        "version": "9.0.2.2",
        "dbname": "sky",
        "dbhost": "127.5.3.2",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "logtimezone": "UTC",
        "installed": true,
        "theme": "",
        "loglevel": 2,
        "maintenance": false
    }
}
```

**Are you using external storage, if yes which one:** No

**Are you using encryption:** No

**Are you using an external user-backend, if yes which one:** No

### Client configuration
**Browser:** Google Chrome

**Operating system:** Ubuntu 16.04

#### Web server error log
```
Nothing unusual here
```

#### ownCloud log (data/owncloud.log)
```
{"reqId":"VkHT3FjMqCcrtOWZ6qDd","remoteAddr":"127.5.3.1","app":"PHP","message":"Undefined offset: 0 at \/var\/lib\/openshift\/572f7fb30c1e6688f2000219\/app-root\/runtime\/repo\/www\/lib\/private\/template\/resourcelocator.php#119","level":3,"time":"2016-05-09T16:39:37+00:00","method":"GET","url":"\/settings\/admin","user":"subins2000"}
{"reqId":"VkHT3FjMqCcrtOWZ6qDd","remoteAddr":"127.5.3.1","app":"vendor","message":"Could not find resource file \"\/davclient.js\/lib\/client.js\"","level":3,"time":"2016-05-09T16:39:37+00:00","method":"GET","url":"\/settings\/admin","user":"subins2000"}
{"reqId":"fJQr0qj30EZpGPZr6Qmt","remoteAddr":"127.5.3.1","app":"PHP","message":"Undefined offset: 0 at \/var\/lib\/openshift\/572f7fb30c1e6688f2000219\/app-root\/runtime\/repo\/www\/lib\/private\/template\/resourcelocator.php#119","level":3,"time":"2016-05-09T16:39:38+00:00","method":"GET","url":"\/apps\/files\/","user":"subins2000"}
{"reqId":"fJQr0qj30EZpGPZr6Qmt","remoteAddr":"127.5.3.1","app":"vendor","message":"Could not find resource file \"\/es6-promise\/dist\/es6-promise.js\"","level":3,"time":"2016-05-09T16:39:38+00:00","method":"GET","url":"\/apps\/files\/","user":"subins2000"}
{"reqId":"fJQr0qj30EZpGPZr6Qmt","remoteAddr":"127.5.3.1","app":"PHP","message":"Undefined offset: 0 at \/var\/lib\/openshift\/572f7fb30c1e6688f2000219\/app-root\/runtime\/repo\/www\/lib\/private\/template\/resourcelocator.php#119","level":3,"time":"2016-05-09T16:39:38+00:00","method":"GET","url":"\/apps\/files\/","user":"subins2000"}
{"reqId":"fJQr0qj30EZpGPZr6Qmt","remoteAddr":"127.5.3.1","app":"vendor","message":"Could not find resource file \"\/davclient.js\/lib\/client.js\"","level":3,"time":"2016-05-09T16:39:38+00:00","method":"GET","url":"\/apps\/files\/","user":"subins2000"}
{"reqId":"BiiRln9\/C+CU\/NcAhSGN","remoteAddr":"127.5.3.1","app":"PHP","message":"Undefined offset: 0 at \/var\/lib\/openshift\/572f7fb30c1e6688f2000219\/app-root\/runtime\/repo\/www\/lib\/private\/template\/resourcelocator.php#119","level":3,"time":"2016-05-09T16:39:41+00:00","method":"GET","url":"\/apps\/files\/","user":"subins2000"}
{"reqId":"BiiRln9\/C+CU\/NcAhSGN","remoteAddr":"127.5.3.1","app":"vendor","message":"Could not find resource file \"\/es6-promise\/dist\/es6-promise.js\"","level":3,"time":"2016-05-09T16:39:41+00:00","method":"GET","url":"\/apps\/files\/","user":"subins2000"}
{"reqId":"BiiRln9\/C+CU\/NcAhSGN","remoteAddr":"127.5.3.1","app":"PHP","message":"Undefined offset: 0 at \/var\/lib\/openshift\/572f7fb30c1e6688f2000219\/app-root\/runtime\/repo\/www\/lib\/private\/template\/resourcelocator.php#119","level":3,"time":"2016-05-09T16:39:41+00:00","method":"GET","url":"\/apps\/files\/","user":"subins2000"}
{"reqId":"BiiRln9\/C+CU\/NcAhSGN","remoteAddr":"127.5.3.1","app":"vendor","message":"Could not find resource file \"\/davclient.js\/lib\/client.js\"","level":3,"time":"2016-05-09T16:39:41+00:00","method":"GET","url":"\/apps\/files\/","user":"subins2000"}
```

**Comment by LukasReschke** [comment]
cc @DeepDiver1975 any thoughts?

**Comment by PVince81** [comment]
Not sure, I thought the public link download had a content-length header.

@lenz1111 did you see this issue while working on the range requests ?

**Comment by lenz1111** [comment]
@PVince81 Public link download had a content-length header - I checked that. Maybe it is problem related to Apache configuration?

**Comment by lenz1111** [comment]
Server (in link) seems to be nginx not apache!

**Comment by Xenopathic** [comment]
```
Could not find resource file \"\/davclient.js\/lib\/client.js\"

                        [core/vendor/davclient.js/lib/client.js] => Array
                        (
                            [expected] => 81d26d67f23736ded76e7ecadd0eb2040820fe65c1c9f57cb348be427d6373c215ecf16fde42a4684d095e01ffd27b14c0342966939896fda2dcad71a87d99cb
                            [current] => 
                        )
```

Please fix this and install all ownCloud files.

**Comment by Xenopathic** [comment]
Gah, I'm not doing well with these GitHub buttons today...

**Comment by DeepDiver1975** [comment]
> Gah, I'm not doing well with these GitHub buttons today...

:question: today :question: :laughing:

**Comment by PVince81** [comment]
Works for me:
```
% curl -D - 'https://owncloud.vincentpetry.net/s/bq1edeIt1tRa4Oz/download' -o worksforme.jpg

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0HTTP/1.1 200 OK
Date: Wed, 11 May 2016 09:06:31 GMT
Server: Apache/2.4.10 (Linux/SUSE)
Strict-Transport-Security: max-age=15768000; includeSubDomains; preload
X-Powered-By: PHP/5.6.1
Set-Cookie: 507a979be894f=115ruc669nchrd6g04slg6fg62dbs9e7rerpghb048t88bh7lh31; path=/; HttpOnly
Expires: 0
Cache-Control: must-revalidate, post-check=0, pre-check=0
Pragma: public
Set-Cookie: oc_sessionPassphrase=hoIOqppp5sIRzdyR61NYqv%2FfbVvzsmiBwJclswLDbDQDIDoVBZyYR7dqgDiK8h5ucVBuE53aLmmcZxW%2BPfyQQ53swd47J%2B3pajMHzYrA3TARlFc5LVvvFGxP4y8o28c8; path=/; secure; httponly
Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src * data: blob:; font-src 'self' data:; media-src *; connect-src *
Content-Disposition: attachment; filename*=UTF-8''worksforme.jpg; filename="worksforme.jpg"
Content-Transfer-Encoding: binary
Content-Length: 115020
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-Robots-Tag: none
X-Frame-Options: SAMEORIGIN
X-Download-Options: noopen
X-Permitted-Cross-Domain-Policies: none
Content-Type: image/jpeg

100  112k  100  112k    0     0  46115      0  0:00:02  0:00:02 --:--:-- 46137
```

This is on 9.0.2.

Maybe check your setup, and/or http://owncloud.org/support for support.

**Comment by subins2000** [comment]
The problem still exists.

I thought this was because of the Apache server. So, I switched to Nginx. But, still the problem exists.

Updated file URL : https://sky-phpgeek.rhcloud.com/index.php/s/qVHNgLfOxeVoLMd/download

**Comment by lenz1111** [comment]
@subins2000 The problem exists because you share folder and trying to download zip archive. Archive file is created on the fly and at the beginning the file length is unknown...

**Comment by subins2000** [comment]
@lenz1111 I'm trying to download an original Zip file. I uploaded ["0.6.zip"](https://sky-phpgeek.rhcloud.com/index.php/s/qVHNgLfOxeVoLMd) file to ownCloud and trying to download it back.

I'm hosting the files of a [software](http://lobby.subinsb.com), so that users can download the software. Without filesize, users tend to cancel the download fearing of high file size.

**Comment by lenz1111** [comment]
@subins2000 Could you try to rename extension for example to: 0.6.dat - and try again?

**Comment by subins2000** [comment]
@lenz1111 The problem exists for .dat files too : https://sky-phpgeek.rhcloud.com/index.php/s/WoBtCBYZCfIuM2w

**Comment by lenz1111** [comment]
@subins2000 tried download from your site - everything is ok - Content-Length header exists...

**Comment by subins2000** [comment]
@lenz1111 That's because I added : 
```php
header("Content-Range: 0-".($fileSize-1)."/".$fileSize);
```
after line 61 in `lib/private/files.php`

When this line is removed, the problem comes back.

**Comment by lenz1111** [comment]
You have added Contant-Range, but there is also Content-Length header... so everything seems to be ok
```
Content-Length:2630147
Content-Range:0-2630146/2630147
```

**Comment by lenz1111** [comment]
@subins2000 Remove your modifications and check if Content-Length header exists or not...

**Comment by Xenopathic** [comment]
@subins2000 That StackOverflow answer seems to suggest you have mod_deflate enabled. This is unsupported as stated in the [documentation](https://doc.owncloud.org/server/9.0/admin_manual/issues/general_troubleshooting.html#web-server-and-php-modules).

@LukasReschke We could really do with that browser-side check you mentioned in #4783

**Comment by subins2000** [comment]
@lenz1111 When I remove my modification, the `Content-Length` header is present, but Chrome doesn't  see it and the download looks like [this](http://imgur.com/6KUcvcNl.png)

But, when I add my modification, Chrome displays the [filesize and time remaining](http://imgur.com/tQxtF9hl.png)

This happens in **Firefox too** : 
```
Without modification - http://imgur.com/3SwdsXTl.png
With modification - http://imgur.com/0pauzXPl.png
```
^ Notice the file size mentioned
@Xenopathic I'm using Nginx now, there is no mod_deflate.

**Comment by subins2000** [comment]
Chromium - Version 49.0.2623.108 Ubuntu 16.04

Firefox - Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:45.0) Gecko/20100101 Firefox/45.0

**Comment by lenz1111** [comment]
@subins2000 If "Content-Length" header exists and it is correct  - that means this issue should be closed...
> Actual behaviour
> The "Content-Length" header is omitted by ownCloud.
> Expected behaviour
> The file should download with "Content-Length" header.

**Comment by subins2000** [comment]
@lenz1111 The issue should have been "File size not showing in browsers"

**Comment by subins2000** [comment]
This bug exists in 9.1 too

**Comment by lenz1111** [comment]
Could you give me shared link to test file on your ownCloud installation?

**Comment by subins2000** [comment]
Hey !

I have applied the fix I suggested in the [comment I made above ^](https://github.com/owncloud/core/issues/24520#issuecomment-218562531) and it works.

Here's the shared link :
```
https://sky-phpgeek.rhcloud.com/index.php/s/BSMk6No0RZ8qZhq/download
```

I have made a tweak in the source code so that the applied fix is unset. To disable the fix I applied, just add GET parameter `lenz-not-apply-fix` to the URL :
```
https://sky-phpgeek.rhcloud.com/index.php/s/BSMk6No0RZ8qZhq/download?lenz-not-apply-fix
```
and the bug will be reproduced.
